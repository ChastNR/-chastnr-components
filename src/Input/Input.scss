.ctrl__cntr {
  input {
    background: none;
    border: none;
    box-sizing: border-box;
    color: #333;
    font-size: 15px;
    font-weight: 400;
    height: 100%;
    letter-spacing: 0;
    line-height: 16px;
    opacity: 0;
    outline: 0;
    padding: 22px 0 9px;
    position: relative;
    text-overflow: ellipsis;
    transition: opacity 0.25s ease;
    white-space: nowrap;
    width: 100%;
    will-change: opacity;
    word-wrap: break-word;

    @media (max-width: 767.98px) {
      font-size: 14px;
    }
  }

  &:focus-within {
    input {
      opacity: 1;
    }
  }

  &.hasValue {
    input {
      opacity: 1;
    }
  }
}
