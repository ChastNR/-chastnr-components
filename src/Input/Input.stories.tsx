/* eslint-disable import/no-extraneous-dependencies */
import { useState } from 'react';
import { ComponentMeta } from '@storybook/react';
import { Input } from './Input';

const componentMeta: ComponentMeta<typeof Input> = {
  title: 'Input',
  component: Input,
  parameters: {
    layout: 'fullscreen',
  },
};

export const DefaultInput = () => {
  const [value, setValue] = useState('');

  return <Input label="Test input" onChange={setValue} value={value} />;
};

export const Form = () => {
  const [value, setValue] = useState('');

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!e.currentTarget.test.value) {
      e.currentTarget.test.focus();
    }
  };

  return (
    <form noValidate onSubmit={handleSubmit}>
      <Input label="Test input" name="test" onChange={setValue} value={value} />
      <button type="submit">Submit</button>
    </form>
  );
};

export default componentMeta;
