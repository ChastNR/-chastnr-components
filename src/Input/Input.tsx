import { useCallback, useRef, useState } from "react";

import { createError } from "../base";
import Control from "../Control";
import { InputTypes } from "../types";

import "./Input.scss";

interface InputProps {
  autoComplete?: "on" | "off";
  className?: string;
  content?: React.ReactNode;
  disabled?: boolean;
  errorMessage?: React.ReactNode;
  label: React.ReactNode;
  maxLength?: number;
  name?: string;
  onBlur?: React.FocusEventHandler<HTMLDivElement>;
  onChange: (value: string) => void;
  onFocus?: React.FocusEventHandler<HTMLDivElement>;
  placeholder?: string;
  regExp?: RegExp;
  requiredMessage?: React.ReactNode;
  startValidate?: boolean;
  type?: InputTypes;
  value: string;
  withSearch?: boolean;
}

const Input: React.FC<InputProps> = ({
  autoComplete = "off",
  children,
  className,
  content,
  disabled,
  errorMessage,
  label,
  maxLength,
  name,
  onBlur,
  onChange,
  onFocus,
  placeholder,
  regExp,
  requiredMessage,
  startValidate = false,
  type = "text",
  value,
}) => {
  const inputRef = useRef<HTMLInputElement>(null);

  const [validate, setValidate] = useState(false);

  const handleFocus = useCallback(
    (e: React.FocusEvent<HTMLDivElement>) => {
      inputRef.current?.focus();

      if (onFocus) {
        onFocus(e);
      }
    },
    [onFocus]
  );

  const handleBlur = useCallback(
    (e: React.FocusEvent<HTMLDivElement>) => {
      if (e.currentTarget.contains(e.relatedTarget as Node) || !inputRef.current) {
        return;
      }

      setValidate(true);

      if (onBlur) {
        onBlur(e);
      }
    },
    [onBlur]
  );

  const handleChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      setValidate(false);
      onChange(e.currentTarget.value);
    },
    [onChange]
  );

  const error = createError({
    validate,
    startValidate,
    requiredMessage,
    regExp,
    value,
    errorMessage,
  });

  return (
    <Control
      className={className}
      content={children}
      disabled={disabled}
      error={error}
      hasValue={!!value}
      label={label}
      name={name}
      onBlur={handleBlur}
      onFocus={handleFocus}
      required={!!requiredMessage}
    >
      <input
        autoComplete={autoComplete}
        disabled={disabled}
        maxLength={maxLength}
        name={name}
        onChange={handleChange}
        placeholder={placeholder}
        ref={inputRef}
        required={!!requiredMessage}
        type={type}
        value={value}
      />
      {content}
    </Control>
  );
};

export default Input;
