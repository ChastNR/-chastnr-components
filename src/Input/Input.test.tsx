import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Input from "./Input";

describe("<Input />", () => {
  const onChangeMock = jest.fn();
  const onBlurMock = jest.fn();
  const onFocusMock = jest.fn();
  const mockSetState = jest.fn();
  const mockUseRef = jest.fn();

  beforeEach(() => {
    jest.mock("react", () => {
      const actualReact = jest.requireActual("react");

      mockUseRef.mockImplementation(actualReact.useRef);

      return {
        ...actualReact,
        useRef: mockUseRef,
        useState: jest.fn().mockImplementation((value) => [value, mockSetState]),
      };
    });
  });

  it("snapshots should be equal", () => {
    const wrapper = shallow(<Input label="Test" onChange={onChangeMock} value="test" />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("props should be equal", () => {
    const label = "Test";
    const name = "testInput";
    const maxLength = 100;
    const type = "password";
    const value = "testValue";
    const placeholder = "testPlaceholder";
    const regExp = new RegExp("pattern");
    const autoComplete = "on";
    const className = "testClassName";
    const errorMessage = "errorMessage";
    const requiredMessage = "requiredMessage";
    const content = "testContent";
    const disabled = true;
    const startValidate = true;

    const input = mount(
      <Input
        autoComplete={autoComplete}
        className={className}
        content={content}
        disabled={disabled}
        errorMessage={errorMessage}
        label={label}
        maxLength={maxLength}
        name={name}
        onChange={onChangeMock}
        placeholder={placeholder}
        regExp={regExp}
        requiredMessage={requiredMessage}
        startValidate={startValidate}
        type={type}
        value={value}
      />
    );

    const {
      className: expectedClassName,
      content: expectedContent,
      disabled: expectedDisabled,
      label: expectedLabel,
      name: expectedName,
      maxLength: expectedMaxLength,
      type: expectedType,
      value: expectedValue,
      placeholder: expectedPlaceholder,
      regExp: expectedRegExp,
      autoComplete: expectedAutoComplete,
      errorMessage: expectedErrorMessage,
      requiredMessage: expectedRequiredMessage,
      startValidate: expectedStartValidate,
    } = input.props();

    expect(expectedClassName).toEqual(className);
    expect(expectedContent).toEqual(content);
    expect(expectedDisabled).toEqual(disabled);
    expect(expectedLabel).toEqual(label);
    expect(expectedName).toEqual(name);
    expect(expectedMaxLength).toEqual(maxLength);
    expect(expectedType).toEqual(type);
    expect(expectedValue).toEqual(value);
    expect(expectedPlaceholder).toEqual(placeholder);
    expect(expectedRegExp).toEqual(regExp);
    expect(expectedAutoComplete).toEqual(autoComplete);
    expect(expectedErrorMessage).toEqual(errorMessage);
    expect(expectedRequiredMessage).toEqual(requiredMessage);
    expect(expectedStartValidate).toEqual(startValidate);
  });

  it("should trigger change event", () => {
    const input = mount(<Input label="Test" onChange={onChangeMock} value="testValue" />);

    input.find("input").simulate("change");
    input.simulate("change", { currentTarget: { value: "test" } });

    expect(onChangeMock).toBeCalledWith("testValue");
  });

  it("should trigger blur event", () => {
    const input = mount(
      <Input label="Test" onBlur={onBlurMock} onChange={onChangeMock} value="testValue" />
    );

    const inputElement = input.find("input");

    inputElement.simulate("focus");
    inputElement.simulate("blur");
    input.simulate("blur");

    expect(onBlurMock).toBeCalled();
  });

  it("should return when blur input node", () => {
    mockUseRef.mockImplementation(() => ({ current: null }));

    const input = mount(<Input label="Test" onChange={onChangeMock} value="testValue" />);

    const inputElement = input.find("input");

    inputElement.simulate("focus");
    inputElement.simulate("blur");
    // input.simulate("blur");

    expect(mockSetState).toBeCalledTimes(0);
  });

  it("should trigger focus event", () => {
    const input = mount(
      <Input label="Test" onChange={onChangeMock} onFocus={onFocusMock} value="testValue" />
    );

    input.find("input").simulate("focus");
    input.simulate("focus");

    expect(onFocusMock).toBeCalled();
  });
});
