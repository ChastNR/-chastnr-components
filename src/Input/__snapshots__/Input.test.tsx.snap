// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> snapshots should be equal 1`] = `
<div
  class="ctrl"
>
  <div
    class="ctrl__cntr testClassName required disabled error hasValue"
    role="textbox"
    tabindex="0"
  >
    <label
      class="ctrl__lb"
      for="testInput"
    >
      Test
    </label>
    <input
      autocomplete="on"
      disabled=""
      maxlength="100"
      name="testInput"
      placeholder="testPlaceholder"
      required=""
      type="password"
      value="testValue"
    />
    testContent
  </div>
  <div
    class="er__cntr"
    style="height: 0px;"
  >
    <div>
      errorMessage
    </div>
  </div>
</div>
`;
